File: modules/benchmarks_external/src/naive_infinite_scroll/scroll_area.dart
library scroll_area;

import 'dart:html';
import 'dart:math' as math;
import 'package:angular/angular.dart';
import 'common.dart';
import 'random_data.dart';

@Component(
    selector: 'scroll-area',
    templateUrl: 'scroll_area.html',
    map: const {'scroll-top': '=>scrollTop',})
class ScrollAreaComponent implements ShadowRootAware {
  Element scrollDiv;
  List<Offering> _fullList;
  List<Offering> visibleItems = [];

  // Init empty maps and populate later. There seems to be a bug in Angular
  // that makes it choke on pre-populated style maps.
  final Map paddingStyle = {};
  final Map innerStyle = {};
  final Map scrollDivStyle = {};

  ScrollAreaComponent() {
    _fullList = generateOfferings(ITEMS);
  }

  @override
  void onShadowRoot(ShadowRoot shadowRoot) {
    scrollDiv = shadowRoot.querySelector('#scrollDiv');
    onScroll();
    scrollDivStyle.addAll({
      'height': '${VIEW_PORT_HEIGHT}px',
      'width': '1000px',
      'border': '1px solid #000',
      'overflow': 'scroll',
    });
    innerStyle['width'] = '${ROW_WIDTH}px';
  }

  set scrollTop(int value) {
    if (value == null || scrollDiv == null) return;
    scrollDiv.scrollTop = value;
  }

  void onScroll() {
    int scrollY = scrollDiv.scrollTop;
    int iStart = scrollY == 0 ? 0 : (scrollY / ITEM_HEIGHT).floor();
    int iEnd = math.min(iStart + VISIBLE_ITEMS + 1, _fullList.length);
    int padding = iStart * ITEM_HEIGHT;
    innerStyle['height'] = '${HEIGHT - padding}px';
    paddingStyle['height'] = '${padding}px';
    visibleItems
      ..clear()
      ..addAll(_fullList.getRange(iStart, iEnd));
  }
}

File: modules/benchmarks/src/naive_infinite_scroll/scroll_area.ts
import {ListWrapper} from '@angular/facade';
import {Math} from '@angular/facade';

import {Component, Directive} from '@angular/core';

import {
  Offering,
  ITEMS,
  ITEM_HEIGHT,
  VISIBLE_ITEMS,
  VIEW_PORT_HEIGHT,
  ROW_WIDTH,
  HEIGHT
} from './common';
import {generateOfferings} from './random_data';
import {ScrollItemComponent} from './scroll_item';
import {NgFor} from '@angular/common';

@Component({
  selector: 'scroll-area',
  directives: [ScrollItemComponent, NgFor],
  template: `
    <div>
        <div id="scrollDiv"
             [style.height.px]="viewPortHeight"
             style="width: 1000px; border: 1px solid #000; overflow: scroll"
             on-scroll="onScroll($event)">
            <div id="padding"></div>
            <div id="inner">
                <scroll-item
                    template="ngFor let item of visibleItems"
                    [offering]="item">
                </scroll-item>
            </div>
        </div>
    </div>`
})
export class ScrollAreaComponent {
  _fullList: Offering[];
  visibleItems: Offering[];

  viewPortHeight: number;
  paddingDiv;
  innerDiv;

  constructor() {
    this._fullList = generateOfferings(ITEMS);
    this.visibleItems = [];
    this.viewPortHeight = VIEW_PORT_HEIGHT;
    this.onScroll(null);
  }

  onScroll(evt) {
    var scrollTop = 0;
    if (evt != null) {
      var scrollDiv = evt.target;
      if (this.paddingDiv == null) {
        this.paddingDiv = scrollDiv.querySelector('#padding');
      }
      if (this.innerDiv == null) {
        this.innerDiv = scrollDiv.querySelector('#inner');
        this.innerDiv.style.setProperty('width', `${ROW_WIDTH}px`);
      }
      scrollTop = scrollDiv.scrollTop;
    }
    var iStart = Math.floor(scrollTop / ITEM_HEIGHT);
    var iEnd = Math.min(iStart + VISIBLE_ITEMS + 1, this._fullList.length);
    var padding = iStart * ITEM_HEIGHT;
    if (this.innerDiv != null) {
      this.innerDiv.style.setProperty('height', `${HEIGHT - padding}px`);
    }
    if (this.paddingDiv != null) {
      this.paddingDiv.style.setProperty('height', `${padding}px`);
    }
    this.visibleItems = ListWrapper.slice(this._fullList, iStart, iEnd);
  }
}
